### PR TITLE
Add utility functions to create polynomial coarsening sequences

### PR DESCRIPTION
**Pull request by peterrum**

### CODE REVIEW

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h [hunk 1]
```diff
@@ -274,6 +327,13 @@ class MGTransferGlobalCoarsening : public dealii::MGTransferBase<VectorType>
     const MGLevelObject<MatrixType> &                         matrices,
     const MGLevelObject<MGTwoLevelTransfer<dim, VectorType>> &transfer);
 
+  /**
+   * TODO.
```
Comment (marcfehling) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:331]: You'll specify this comment and some other parts in #11699. So this PR is blocked by #11699?

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h [hunk 2]
```diff
@@ -551,8 +551,7 @@ namespace internal
             for (auto cell_id : i.second)
               {
                 typename MeshType::cell_iterator cell(
-                  *tria_dst.create_cell_iterator(
-                    cell_id_translator.to_cell_id(cell_id)),
```
Comment (marcfehling) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:555]: CellId::to_cell is deprecated. I think you can safely keep the old code here and in all the other occasions.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h [hunk 3]
```diff
@@ -581,8 +580,8 @@ namespace internal
           {
             const auto cell_id = cell_id_translator.to_cell_id(id);
 
-            typename MeshType::cell_iterator cell_(
-              *tria_dst.create_cell_iterator(cell_id), &dof_handler_dst);
```
Comment (marcfehling) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:585]: Same here.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h [hunk 4]
```diff
@@ -696,8 +695,7 @@ namespace internal
           if (is_cell_locally_owned)
             {
               (typename MeshType::cell_iterator(
-                 *mesh_fine.get_triangulation().create_cell_iterator(
-                   cell->id()),
```
Comment (marcfehling) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:700]: Same here.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h [hunk 5]
```diff
@@ -734,8 +732,7 @@ namespace internal
...
```
Comment (marcfehling) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:738]: Same here.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h [hunk 6]
```diff
@@ -1620,6 +1617,57 @@ namespace internal
...
+
+
+
+  std::vector<unsigned int>
+  create_polynomial_coarsening_sequence(
+    const unsigned int                      max_degree,
+    const PolynomialCoarseningSequenceType &p_sequence)
+  {
+    std::vector<unsigned int> degrees;
+    degrees.push_back(max_degree);
```
Comment (marcfehling) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:1650]:
```suggestion
    std::vector<unsigned int> degrees = {max_degree};
```

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h [hunk 7]
```diff
@@ -37,6 +37,59 @@ namespace internal
...
+   * determine the next coarser degree.
+   */
+  unsigned int
+  create_next_polynomial_coarsening_degree(
+    const unsigned int                      degree,
+    const PolynomialCoarseningSequenceType &p_sequence);
+
+  /**
+   * For a given @p max_degree and polynomial coarsening sequence @p p_sequence,
+   * determine the full sequence of polynomial degrees.
```
Comment (marcfehling) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:82]:
```suggestion
   * determine the full sequence of polynomial degrees in ascending order.
```

**Review by marcfehling — APPROVED** [review]

**Comment by peterrum** [comment]
@marcfehling Thanks for the review. I have removed the accidental copy-and-paste stuff.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h [hunk 8]
```diff
@@ -37,6 +37,59 @@ namespace internal
...
+   * Common polynomial coarsening sequences.
+   *
+   * @note These polynomial coarsening sequences up to a degree of 9 are
+   *   precompiled in MGTwoLevelTransfer. See also:
+   *   MGTwoLevelTransfer::fast_polynomial_transfer_supported()
+   */
+  enum class PolynomialCoarseningSequenceType
+  {
+    /**
+     * Half polynomial degree. E.g., for degree=7 following sequence would
```
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:55]:
```suggestion
     * Half polynomial degree by integer division. For example, for degree=7 the following sequence would be obtained:
```

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h [hunk 9]
```diff
@@ -37,6 +37,59 @@ namespace internal
...
+   * determine the next coarser degree.
+   */
+  unsigned int
+  create_next_polynomial_coarsening_degree(
+    const unsigned int                      degree,
+    const PolynomialCoarseningSequenceType &p_sequence);
+
+  /**
+   * For a given @p max_degree and polynomial coarsening sequence @p p_sequence,
+   * determine the full sequence of polynomial degrees in ascending order.
```
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:82]:
```suggestion
   * determine the full sequence of polynomial degrees, sorted in ascending order.
```

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h [hunk 10]
```diff
@@ -1620,6 +1620,56 @@ namespace internal
...
+    unsigned int previous_fe_degree = max_degree;
+    while (previous_fe_degree > 1)
+      {
+        const unsigned int level_degree =
+          create_next_polynomial_coarsening_degree(previous_fe_degree,
+                                                   p_sequence);
+
+        degrees.push_back(level_degree);
+        previous_fe_degree = level_degree;
+      }
```
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:1663]: Why not simply
```suggestion
    unsigned int previous_fe_degree = max_degree;
    while (degrees.back() > 1)
      degrees.push_back(create_next_polynomial_coarsening_degree(degrees.back(),
                                                                 p_sequence));
```

**Review by kronbichler — CHANGES_REQUESTED** [review]

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h (hunk 10 above)
Comment (peterrum) [reply include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h]: Much better!

**Comment by peterrum** [comment]
@kronbichler Thanks for the review. I have addressed the comments.

**Review by kronbichler — APPROVED** [review]

**Pull request merged**